src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -0,0 +1,345 @@
+                         << worker->GetNodeID() << ", address " << worker->GetIpAddress()
+                         << ":" << worker->GetPort() << " as session is broken.";
+        if (auto gcs_node_manage = gcs_node_manager_.lock()) {
+          if (auto node = gcs_node_manage->GetNode(worker->GetNodeID())) {
+            // TODO(zsl): Treat it as WorkerFailure when !status.ok(), it may be a little
+            // strict too. Maybe better to retry creating actor on this worker.
+            node->OnWorkerFailure(worker->GetWorkerID());
+          }
+        }
+        return;
+      }
+
+      RAY_LOG(INFO) << "Succeed in creating actor " << actor->GetActorID()
+                    << " on worker " << worker->GetWorkerID() << " at node "
+                    << actor->GetNodeID();
+      OnActorCreateSuccess(actor, worker);
+    }
+  });
+}
+
+void GcsActorManager::FailureActor(std::shared_ptr<GcsActor> actor) {
+  RAY_CHECK(actor != nullptr);
+  auto node_id = actor->GetNodeID();
+  auto worker_id = actor->GetWorkerID();
+  RAY_LOG(WARNING) << "Failure actor " << actor->GetActorID() << " on worker "
+                   << worker_id << " at node " << node_id;
+  actor->SetNodeID(ClientID::Nil());
+
+  auto actor_table_data = actor->GetMutableActorTableData();
+  auto remaining_reconstructions = actor_table_data->remaining_reconstructions();
+  if (remaining_reconstructions > 0) {
+    actor_table_data->set_remaining_reconstructions(--remaining_reconstructions);
+    actor_table_data->set_state(rpc::ActorTableData::RECONSTRUCTING);
+    AsyncFlush(actor, [this, actor] { Enqueue(actor); });
+  } else {
+    actor_table_data->set_state(rpc::ActorTableData::DEAD);
+    AsyncFlush(actor, nullptr);
+  }
+}
+
+void GcsActorManager::FailureAllActorsAssociateWithNode(std::shared_ptr<GcsNode> node) {
+  for (auto &entry : actor_indexer_) {
+    if (entry.second->GetNodeID() == node->GetNodeId()) {
+      FailureActor(entry.second);
+    }
+  }
+}
+
+void GcsActorManager::OnActorCreateSuccess(std::shared_ptr<GcsActor> actor,
+                                           std::shared_ptr<GcsLeasedWorker> worker) {
+  RAY_CHECK(actor && worker && actor->GetWorker());
+  actor->SetState(rpc::ActorTableData::ALIVE);
+  AsyncFlush(actor, nullptr);
+}
+
+void GcsActorManager::Enqueue(std::shared_ptr<GcsActor> actor) {
+  RAY_CHECK(actor->GetWorker() == nullptr);
+  io_context_.post([this, actor] { actors_.push(actor); });
+}
+
+void GcsActorManager::RegisterActor(const ray::rpc::CreateActorRequest &request,
+                                    std::function<void(Status)> &&callback) {
+  RAY_CHECK(callback);
+  const auto &actor_creation_task_spec = request.task_spec().actor_creation_task_spec();
+  auto actor_id = ActorID::FromBinary(actor_creation_task_spec.actor_id());
+  RAY_LOG(INFO) << "Try register actor " << actor_id;
+
+  if (actor_indexer_.count(actor_id) != 0) {
+    // The actor is already exist, just ignore it.
```

Comment:
> When does this situation come up? It doesn't seem okay to ignore a duplicate actor creation request.

Actually, this logic is similar to the logic on the worker side.

Consider that if some of actors are created already in cluster
and went well. Suddenly, the GCS Server is crashed. The following steps will be:

1. GCS Server will restart and load all actor_table_data from storage and try reconstruct all of actors.
2. GCS Server will try to lease worker from the node the actor previously tied to.
3. GCS Server will try to create the actor on the leased worker.

If the actor is already created on the worker in current raylet, then the raylet should fast reply the received LeaseWorker request come from GCS Server just and lease nothing because that the actor is already created on a worker in this node before the GCS Server restart.

Reply:
It seems that this function is called when a worker invokes `Actor.remote`, right? The scenario you explained above doesn't seem to explain why we would receive duplicate `CreateActor` requests.

Reply:
> It seems that this function is called when a worker invokes `Actor.remote`, right? The scenario you explained above doesn't seem to explain why we would receive duplicate `CreateActor` requests.

Sorry, I read the wrong code before.

Consider that if Driver/Worker sends the `CreateActorRequest` to the GCS Server, and the network has some problem before the Driver/Worker received the `CreateActorReply`, then the Driver/Worker will retry creating actor, which will cause that the GCS Server receives duplicate `CreateActor` requests.

If the actor is already records on the GCS Server side, the GCS Server will be responsible for creating the actor regardless of whether the Driver/Worker sends the request to create the actor again, so we just need fast reply OK to the Driver/Worker that the actor is already recored by GCS Server.